modules/gdnative/pluginscript/pluginscript_language.cpp

```diff
@@ -0,0 +1,407 @@
+#include <stdlib.h>
+// Godot imports
+#include "core/os/file_access.h"
+#include "core/os/os.h"
+#include "core/project_settings.h"
+// PluginScript imports
+#include "pluginscript_language.h"
+#include "pluginscript_script.h"
+
+String PluginScriptLanguage::get_name() const {
+	return String(this->_desc.name);
+}
+
+void PluginScriptLanguage::init() {
+	return this->_desc.init();
+}
+
+String PluginScriptLanguage::get_type() const {
+	return String(this->_desc.type);
+}
+
+String PluginScriptLanguage::get_extension() const {
+	return String(this->_desc.extension);
+}
+
+Error PluginScriptLanguage::execute_file(const String &p_path) {
+	// TODO: pretty sure this method is totally deprecated and should be removed...
+	return OK;
+}
+
+void PluginScriptLanguage::finish() {
+	return this->_desc.finish();
+}
+
+/* EDITOR FUNCTIONS */
+
+void PluginScriptLanguage::get_reserved_words(List<String> *p_words) const {
+	if (this->_desc.reserved_words) {
+		const char **w = this->_desc.reserved_words;
+		while (*w) {
+			p_words->push_back(*w);
+			w++;
+		}
+	}
+}
+
+void PluginScriptLanguage::get_comment_delimiters(List<String> *p_delimiters) const {
+	if (this->_desc.comment_delimiters) {
+		const char **w = this->_desc.comment_delimiters;
+		while (*w) {
+			p_delimiters->push_back(*w);
+			w++;
+		}
+	}
+}
+
+void PluginScriptLanguage::get_string_delimiters(List<String> *p_delimiters) const {
+	if (this->_desc.string_delimiters) {
+		const char **w = this->_desc.string_delimiters;
+		while (*w) {
+			p_delimiters->push_back(*w);
+			w++;
+		}
+	}
+}
+
+Ref<Script> PluginScriptLanguage::get_template(const String &p_class_name, const String &p_base_class_name) const {
+	Script *ns = this->create_script();
+	Ref<Script> script = Ref<Script>(ns);
+	if (this->_desc.get_template_source_code) {
+		godot_string src = this->_desc.get_template_source_code((godot_string *)&p_class_name, (godot_string *)&p_base_class_name);
+		script->set_source_code(*(String *)&src);
+	}
+	return script;
+}
+
+bool PluginScriptLanguage::validate(const String &p_script, int &r_line_error, int &r_col_error, String &r_test_error, const String &p_path, List<String> *r_functions) const {
+	PoolStringArray functions;
+	if (this->_desc.validate) {
+		bool ret = this->_desc.validate(
+				(godot_string *)&p_script,
+				&r_line_error,
+				&r_col_error,
+				(godot_string *)&r_test_error,
+				(godot_string *)&p_path,
+				(godot_pool_string_array *)&functions);
+		for (int i = 0; i < functions.size(); i++) {
+			r_functions->push_back(functions[i]);
+		}
+		return ret;
+	}
+	return true;
+}
+
+Script *PluginScriptLanguage::create_script() const {
+	PluginScript *script = memnew(PluginScript());
+	// I'm hurting kittens doing this I guess...
+	script->init(const_cast<PluginScriptLanguage *>(this));
+	return script;
+}
+
+bool PluginScriptLanguage::has_named_classes() const {
+	// TODO? GDScript just returns false...
+	return false;
```

Comment:
This should actually be configurable by a scripting language. GDScript and VisualScript have it as false, while NativeScript and Mono have it as true.

It really depends on the language, so I would definitely make this configurable.

Reply:
done.